spring-session/src/main/java/org/springframework/session/context/SessionContextHolder.java

```diff
@@ -0,0 +1,162 @@
+ * <p>
+ * To specify which strategy should be used, you must provide a mode setting. A mode setting is one of the three valid <code>MODE_</code> settings defined as
+ * <code>static final</code> fields, or a fully qualified classname to a concrete implementation of {@link SessionContextHolderStrategy} that provides a public
+ * no-argument constructor.
+ * <p>
+ * There are two ways to specify the desired strategy mode <code>String</code>. The first is to specify it via the system property keyed on
+ * {@link #SYSTEM_PROPERTY}. The second is to call {@link #setStrategyName(String)} before using the class. If neither approach is used, the class will default
+ * to using {@link #MODE_THREADLOCAL}, which is backwards compatible, has fewer JVM incompatibilities and is appropriate on servers (whereas
+ * {@link #MODE_GLOBAL} is definitely inappropriate for server use).
+ *
+ * @author Francisco Spaeth
+ * @since 1.1
+ * 
+ */
+public class SessionContextHolder {
+
+	public static final String MODE_THREADLOCAL = "MODE_THREADLOCAL";
+	public static final String MODE_INHERITABLETHREADLOCAL = "MODE_INHERITABLETHREADLOCAL";
+	public static final String MODE_GLOBAL = "MODE_GLOBAL";
+	public static final String SYSTEM_PROPERTY = "spring.session.strategy";
+	private static String strategyName = System.getProperty(SYSTEM_PROPERTY);
+	private static SessionContextHolderStrategy strategy;
+	private static int initializeCount = 0;
+
+	static {
+		initialize();
+	}
+
+	/**
+	 * Explicitly clears the context value.
+	 */
+	public static void clearContext() {
+		strategy.clearContext();
+	}
+
+	/**
+	 * Obtain the current <code>Session</code>.
+	 * 
+	 * @return the security context, <code>null</code> when no session available
+	 */
+	public static Session getSession() {
+		SessionContext ctx = strategy.getContext();
+
+		if (ctx == null) {
+			return null;
+		}
+
+		return ctx.getSession();
+	}
+
+	/**
+	 * Primarily for troubleshooting purposes, this method shows how many times the class has re-initialized its {@link SessionContextHolderStrategy}.
+	 * 
+	 * @return the count (should be one unless you've called {@link #setStrategyName(String)} to switch to an alternate strategy).
+	 */
+	public static int getInitializeCount() {
+		return initializeCount;
+	}
+
+	private static void initialize() {
+		if (StringUtils.isEmpty(strategyName)) {
+			strategyName = MODE_THREADLOCAL;
+		}
+
+		if (MODE_THREADLOCAL.equals(strategyName)) {
+			strategy = new ThreadLocalSessionContextHolderStrategy();
+		} else if (MODE_INHERITABLETHREADLOCAL.equals(strategyName)) {
+			strategy = new InheritableThreadLocalSessionContextHolderStrategy();
+		} else if (MODE_GLOBAL.equals(strategyName)) {
+			strategy = new GlobalSessionContextHolderStrategy();
+		} else {
+			instantiateCustomStrategy();
+		}
+
+		initializeCount++;
+	}
+
+	private static void instantiateCustomStrategy() {
+		try {
+			Class<?> clazz = Class.forName(strategyName);
```

Comment:
I foresee OSGi users crying because of this but this is something normal for them :-)